### PR TITLE
[SM][com-160] Added input fix for press, hold, & release

### PR DESCRIPTION
**Pull request by Sepehrman**

### CODE REVIEW

File: .idea/.idea.DKEngine/.idea/indexLayout.xml
Comment (le-rich) [inline .idea/.idea.DKEngine/.idea/indexLayout.xml:1]: .idea folder should not be included

File: Core/include/Input.h [hunk 1]
```diff
@@ -1,87 +1,82 @@
-#pragma once
-#include <unordered_map>
-#include <glad/glad.h> // Include glad first
+#include <queue>
+#include <glad/glad.h>
 #include <GLFW/glfw3.h>
 #include <iostream>
```
Comment (le-rich) [inline Core/include/Input.h:4]: Nitpick: stdlib includes should go after ours.

File: Core/include/Input.h [hunk 2]
```diff
@@ -1,87 +1,82 @@
-#pragma once
```
Comment (le-rich) [inline Core/include/Input.h:1]: You removed #pragma once.

File: Core/include/Input.h [hunk 3]
```diff
@@ -1,87 +1,82 @@
...
 class Input
 {
 public:
-    // static maps that keep track of the current state of keypresses & mouse clicks (pressed or unpress)
-    static std::unordered_map<int, bool> keys;
-    static std::unordered_map<int, bool> mouseButtons;
+    enum ActionType { PRESS, RELEASE, HOLD };
 
+    struct InputEvent {
+        int code;
```
Comment (le-rich) [inline Core/include/Input.h:15]: You call it code here but you assign it a key, is there a reason you call it code?

File: Core/include/Input.h [hunk 4]
```diff
@@ -1,87 +1,82 @@
...
-            std::cout << "Left mouse button clicked" << std::endl;
-        }
-
-        if (IsMouseButtonPressed(GLFW_MOUSE_BUTTON_RIGHT))
-        {
-            std::cout << "Right mouse button clicked" << std::endl;
+            InputEvent event = eventQueue.front();
+            eventQueue.pop();
+
+            // Chekcs for press, release or hold
```
Comment (le-rich) [inline Core/include/Input.h:43]: Typo in comment

**Review by le-rich — CHANGES_REQUESTED** [review]

File: Core/include/Input.h (hunk 3 above)
Comment (Sepehrman) [reply Core/include/Input.h:15]: it's keycode basically
sorrt for the confusion

**Review by Tissure — CHANGES_REQUESTED** [review]

File: Core/include/Input.h [hunk 5]
```diff
@@ -1,87 +1,82 @@
...
-    static bool IsKeyPressed(int key)
-    {
-        return keys[key];
+        ActionType actionType = (action == GLFW_PRESS) ? PRESS : (action == GLFW_RELEASE ? RELEASE : HOLD);
+        eventQueue.push({ button, actionType });
     }
-
-    static bool IsMouseButtonPressed(int button)
+    
+    static void RunInputListener(AE86::RigidBody* rigidBody)
```
Comment (Tissure) [inline Core/include/Input.h:36]: How will we handle additional components that use input?
Should be converted to a switch-case for faster parsing

**Review by Tissure — APPROVED** [review]
Comments should be addressed and ticketed.

**Review by le-rich — APPROVED** [review]

**Pull request merged**